Sources/PackageGraph/AvailableModules.swift

```diff
@@ -0,0 +1,133 @@
+        case target(name: String)
+        case product(name: String, package: String?)
+
+        static func <(lhs: TargetDependency, rhs: TargetDependency) -> Bool {
+            switch (lhs, rhs) {
+            case (.target(name: let lhsName), .target(name: let rhsName)):
+                lhsName < rhsName
+            case (.product(name: let lhsName, package: nil),
+                  .product(name: let rhsName, package: nil)):
+                lhsName < rhsName
+            case (.product(name: _, package: nil),
+                  .product(name: _, package: _?)):
+                true
+            case (.product(name: _, package: _?),
+                  .product(name: _, package: nil)):
+                false
+            case (.product(name: let lhsName, package: let lhsPackage?),
+                  .product(name: let rhsName, package: let rhsPackage?)):
+                (lhsPackage, lhsName) < (rhsPackage, rhsName)
+            case (.product, .target):
+                false
+            case (.target, .product):
+                true
+            }
+        }
+    }
+
+    /// The set of modules that are available within the package described by
+    /// the manifest, along with the target dependency required to reference
+    /// the module.
+    package var modules: [String: TargetDependency] = [:]
+}
+
+extension ModulesGraph {
+    /// A flat list of available modules, used as an intermediary for the
+    /// creation of an `AvailableModules` instance.
+    fileprivate typealias AvailableModulesList =
+        [(String, AvailableModules.TargetDependency)]
+
+    /// Collect the module names that are made available by all of the products
+    /// in this package, along with how the target dependency should be
+    /// expressed to make the corresponding modules importable.
+    ///
+    /// The resulting module names are available to any package with a
+    /// dependency on this package.
+    fileprivate func productsAsAvailableModules(
+        from package: ResolvedPackage
+    ) -> AvailableModulesList {
+        package.products.flatMap { product in
+            let productDependency: AvailableModules.TargetDependency = .product(
+                name: product.name,
+                package: product.packageIdentity.description
+            )
+
+            return product.targets.map { target in
+                (target.c99name, productDependency)
+            }
+        }
+    }
+
+    /// Collect the module names that are made available by all of the targets
+    /// in this package, along with how the target dependency should be
+    /// expressed to make the corresponding module importable.
+    ///
+    /// The resulting module names are available within the targets of this
+    /// package.
+    fileprivate func targetsAsAvailableModules(
+        in package: ResolvedPackage
+    ) -> AvailableModulesList {
+        package.targets.map { target in
+            (target.c99name, .target(name: target.name))
```

Comment:
I don't think this is guaranteed to be correct for C targets or binary targets, and it should exclude any plugin and macro targets

Reply:
For binary targets we at least tell people to match the target name and module name, but we aren't verifying it.